[WIP] add support for spock tests

jarlehansen opened the pull request:
This is a work-in-progress branch to add Spock support to fluentlenium. 

The spock adapter uses the majority of the code from the junit adapter. Spock supports junit @Rule, so the amount of code/logic needed to set this up is very limited. The one thing I had to solve is that both spock and fluentlenium expects to use inheritance (Specification and FluentTest). I did this by creating a FluentSpecification class that extends the normal spock Specification and also contains the @Rule's.

The one downside currently is that you have to use a variable to get a hold of the fluentlenium methods (currently it is called 'fl'). There might be ways to solve this as well? If you have any suggestions, please let me know.

I have not added examples yet to the project, but I do have them in a separate project. I will add them when we are closer to getting this branch merged, I just wanted to get this into a pull request so you can take a look.

Here is an example of using the fluentlenium-spock adapter:

```
@Wait
class FluentleniumSpec extends FluentSpecification {

    def "Title of duck duck go"() {
        given:
        fl.goTo("https://duckduckgo.com")

        when:
        fl.$("#search_form_input_homepage").fill().with("FluentLenium")
        fl.$("#search_button_homepage").submit()

        then:
        fl.window().title().contains("FluentLenium")
    }
}
```

Toilal commented:
Interesting, I didn't heard of spoke before. Does it require some compiler plugin in maven to make this compile ?

Toilal commented:
Ho, it's groovy :)

jarlehansen commented:
It is groovy, so I am using this: gmavenplus-plugin

Toilal commented:
We are open to contributions as much as possible, especially to support every possible test runner :) thanks

filipcynarski commented:
I know spock, it is really good library to write tests on BDD style. It would be great if we can extend FluentLenium to support Spock.

filipcynarski commented:
@jarlehansen please give ma a day or two to review these changes. One more request, could you please extend documentation and add tests (unit, integration) for this module?

jarlehansen commented:
@filipcynarski yes, I will add some tests and examples.
Just wanted to get this submitted as work-in-progress for now, but I plan on extending it.

filipcynarski commented on fluentlenium-spock/src/main/groovy/org/fluentlenium/adapter/spock/FluentSpecification.groovy:
```diff
@@ -0,0 +1,56 @@
+
+class FluentSpecification extends Specification {
+
+    protected FluentSpecificationAdapter fl = new FluentSpecificationAdapter()
```
Could you please rename fl to something more meaningful

Toilal replied:
Well ... I even think this variable should be private and annotated with @Delegate (Lombok) for this to simulate the multiple inheritance, so user could then write this : 

```groovy
@Wait
class FluentleniumSpec extends FluentSpecification {

    def "Title of duck duck go"() {
        given:
        goTo("https://duckduckgo.com")

        when:
        $("#search_form_input_homepage").fill().with("FluentLenium")
        $("#search_button_homepage").submit()

        then:
        window().title().contains("FluentLenium")
    }
}
```

jarlehansen replied:
I checked the project lombok page, and it seems like the @Delegate is move to experimental and they actually do not recommend using it anymore.

I did add a groovy delegate and it does seem to work. I have added a new commit with this change to the same PR.

jarlehansen commented:
I just added a spock example and started on the integration tests. All feedback is appreciated, particularly I am not sure if I have added the structure of the integration tests correctly.

I looked at how the junit adapter was setup, and one question I have is about this:
```
<profiles>
        <profile>
            <id>integration-tests-dev</id>
            <properties>
                <it.project.version>3.0.0-SNAPSHOT</it.project.version>
            </properties>
        </profile>
    </profiles>
```

How is this actually triggered? Is this related to the `maven-invoker-plugin`?
And, why is the version here 3.0.0-SNAPSHOT? Shouldnt it be 3.1.2-SNAPSHOT?

filipcynarski commented:
@jarlehansen it would be easier for me to review your changes in IDE, Ive created feature/spock-support branch where you can make pull requests. It should be easier for me to maintain that feature before merging it to develop branch.

jarlehansen commented:
@filipcynarski ok, thanks. will create a new pull request
